### ISSUE TITLE
Finetuning Citrinet using speech_to_text_bpe poor WER

### ISSUE THREAD

**Issue by peterhanlon** [NVIDIA/NeMo#2665]
**Describe the bug**

I'm trying to use speech_to_text_bpe.py to fine-train Citrinet. I've managed to fine-train citrinet with reasonable success using my own code getting a WER of 16 on 8Khz audio upsampled to 16Khz. However, I would rather use the scripts if possible as it will be easier to maintain longer term. When I use the speech_to_text_xxx.py scripts, the results are poor with my WER sitting at around 95, so clearly I'm doing something wrong. 

**Steps/Code to reproduce bug**

I first create by subword tokenizer, this was using spe but i've also tried wpe. To be honest I'm not sure about the correct setting for vocab_size and if 'spe' is the correct tokenizer to use.

!python /root/asr16000/NeMo/scripts/tokenizers/process_asr_text_tokenizer.py \
  --manifest="/root/asr16000/training/training_samples.json" \
  --data_root="/root/asr16000/tokenizer" \
  --vocab_size=64 \
  --tokenizer='spe' \
  --no_lower_case \
  --spe_type='unigram' \
  --spe_character_coverage=1.0 \
  --log

Then I train the model, I assume I should use the '_bpe' script version here.

!python /root/asr16000/NeMo/examples/asr/speech_to_text_bpe.py \
    --config-path='/root/asr16000/NeMo/examples/asr/conf/citrinet/' \
    --config-name='citrinet_512' \
    model.train_ds.manifest_filepath='/root/asr16000/training/training_samples.json' \
    model.train_ds.batch_size=16 \
    model.validation_ds.manifest_filepath='/root/asr16000/training/eval_samples.json' \
    model.validation_ds.batch_size=16 \
    model.test_ds.manifest_filepath='/root/asr16000/training/eval_samples.json' \
    model.test_ds.batch_size=16 \
    model.tokenizer.dir='/root/asr16000/tokenizer/tokenizer_spe_unigram_v64/' \
    model.tokenizer.type='bpe' \
    model.optim.name="adamw" \
    model.optim.lr=0.001 \
    model.optim.betas=[0.9,0.999] \
    model.optim.weight_decay=0.0001 \
    model.optim.sched.warmup_steps=2000 \
    trainer.gpus=1 \
    trainer.accelerator="ddp" \
    trainer.max_epochs=100 \
    exp_manager.create_wandb_logger=True \
    exp_manager.wandb_logger_kwargs.name="nemo-asr-script-citrinet" \
    exp_manager.wandb_logger_kwargs.project="speech_to_text" 

The /root/asr16000/NeMo directory is a clone of the 1.2 nemo repo. I have about 350 hours of audio for training.

**Expected behavior**

I would expect the WER to be better than 95, I was hoping for something lower than 16. Any pointers would be much appreciated.

**Environment overview (please complete the following information)**

 - Environment location: AWS Sage Maker
 - Method of NeMo install: %pip install nemo_toolkit['all']==1.2.0

**Environment details**

- OS version : Debian GNU/Linux 10 (buster)
- PyTorch version: 1.9.0+cu102
- Python version: 3.6.13

**Additional context**
GPU model A100

**Comment by titu1994** [comment]
The example scripts in Nemo are just that - examples. You are encouraged to write your own scripts based on ideas from that example. 

Tokekizer building is correct, but vocab size is not. Since it is a CTC model, you can refer to the ASR CTC Finetuning tutorial - there is a snippet there to check that samples from the dataset are not being wasted due to CTC loss limitations. 

For a good default, 1024 SPE/BPE tokens is generally recommended. You can go smaller, but then you just make sure that transcripts are not being skipped, I fear that is primarily the reason the model fails to train. 

> Then I train the model, I assume I should use the '_bpe' script version here.

_bpe.py scripts are for any type of subword model - either WPE or SPE or any other tokenizer in the future. 

The script is nothing special, whatever scores you were able to get on your personal workflow should be possible with the script too, so switching to it alone should not make any difference. 

One thing you'll note- the scripts are for "from scratch" training by default. This is not generally the best course of action. Even for different domains, or even different languages, starting from a pretrained Checkpoint is significantly better. 

Since you are using the Citrinet 512 model, you can actually add one line to your config to force the model to load weights from a pretrained checkpoint (iff you use vocab size 1024 in your tokenizer) - 

`+init_from_pretrained_model="stt_en_citrinet_512"`

This will build your model and then load the checkpoint of that model into your custom model. This will significantly speed up convergence, since the original checkpoint was trained on roughly 7000 hours of speech.

**Comment by peterhanlon** [comment]
Thanks for the fast feedback! I meant to include the init_from_pretrained_model in the example I pasted above. I've tried a number of different parameter combinations including the +init_from_pretrained_model="stt_en_citrinet_512" without any success. I'm going to try with 1024 and see how that goes. For my use-case, I'm training on 8Khz audio that has been upsampled to 16khz. The audio is clear, would you recommend citrinet or conformer to fine train against? Thanks for all your help

**Comment by titu1994** [comment]
Citrinet and Conformer will both get similar results so you could try either. Could you describe your custom code that was able to achieve 16 WER?

**Comment by peterhanlon** [comment]
The code isn't particularly unique, I based it on the example https://github.com/NVIDIA/NeMo/blob/v1.1.0/tutorials/asr/08_ASR_with_Subword_Tokenization.ipynb. I don't have it in front of me at the moment but the learning rate was 0.001, I was using a vocab size of 64, using wpe and spe_type of unigram, batch size of 16 fine tuned from 'stt_en_citrinet_512'.

**Comment by peterhanlon** [comment]
I was using the same training data as I'm using now. I's 8Khz but I upsampled it using sox to 16khz. The audio is pretty clear for 8Khz recordings.

**Comment by titu1994** [comment]
That seems reasonable way to train. I would suggest going through the notebook and checking if the steps are being replicated in the script.
